// metro/src/node-haste/DependencyGraph/ResolutionRequest.js
/**
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */

'use strict';

const ModuleResolution = require('./ModuleResolution');

const isAbsolutePath = require('absolute-path');
const path = require('path');const

DuplicateHasteCandidatesError = require('jest-haste-map').ModuleMap.DuplicateHasteCandidatesError;const






isRelativeImport = ModuleResolution.isRelativeImport;






































class ResolutionRequest {




  constructor(options) {
    this._options = options;
    this._resetResolutionCache();
  }

  resolveDependency(fromModule, toModuleName) {
    const resHash = getResolutionCacheKey(fromModule.path, toModuleName);

    const immediateResolution = this._immediateResolutionCache[resHash];
    if (immediateResolution) {
      return immediateResolution;
    }

    const cacheResult = result => {
      this._immediateResolutionCache[resHash] = result;
      return result;
    };

    const resolver = this._options.moduleResolver;
    const platform = this._options.platform;

    if (
    isAbsolutePath(toModuleName) ||
    isRelativeImport(toModuleName) ||
    this._options.helpers.isNodeModulesDir(fromModule.path))
    {
      return cacheResult(
      resolver.resolveNodeDependency(fromModule, toModuleName, platform));

    }

    return cacheResult(
    ModuleResolution.tryResolveSync(
    () => this._resolveHasteDependency(fromModule, toModuleName, platform),
    () =>
    resolver.resolveNodeDependency(fromModule, toModuleName, platform)));


  }

  _resolveHasteDependency(
  fromModule,
  toModuleName,
  platform)
  {
    const rs = this._options.moduleResolver;
    try {
      return rs.resolveHasteDependency(fromModule, toModuleName, platform);
    } catch (error) {
      if (error instanceof DuplicateHasteCandidatesError) {
        throw new AmbiguousModuleResolutionError(fromModule.path, error);
      }
      throw error;
    }
  }

  _resetResolutionCache() {
    this._immediateResolutionCache = Object.create(null);
  }

  getResolutionCache() {
    return this._immediateResolutionCache;
  }}


function getResolutionCacheKey(modulePath, depName) {
  return `${path.resolve(modulePath)}:${depName}`;
}

class AmbiguousModuleResolutionError extends Error {



  constructor(
  fromModulePath,
  hasteError)
  {
    super(
    `Ambiguous module resolution from \`${fromModulePath}\`: ` +
    hasteError.message);

    this.fromModulePath = fromModulePath;
    this.hasteError = hasteError;
  }}


ResolutionRequest.AmbiguousModuleResolutionError = AmbiguousModuleResolutionError;

module.exports = ResolutionRequest;
